### PR TITLE
Fixes #33690 - Unify toast notification logic in one component

### PR DESCRIPTION
**Pull request by tbrisker**
Instead of having the actions and reducers separate from the actual
component, this refactor leverages the slice pattern to unify all of the
logic and remove a lot of boilerplate.

~The second commit adds a hook that can be used to add a notification.~

### CODE REVIEW

**Comment by tbrisker** [comment]
[test foreman]

File: webpack/assets/javascripts/react_app/components/ToastsList/hooks.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { useDispatch } from 'react-redux';
+import { addToast } from './slice';
+
+/**
+ * Hook for adding a toast
+ * @returns function(message) that dispatches the action to add the message to the toast list
+ */
+export const useToast = () => {
```
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/components/ToastsList/hooks.js]: TBH I'm not sure it makes sense to add a hook for this, might be better to use directly? I've split it into a second commit so it's easy to drop, would love another opinion about this.

Comment (tbrisker) [reply webpack/assets/javascripts/react_app/components/ToastsList/hooks.js]: Looks like the hook didn't work quite as expected, dropped it for now.

**Comment by tbrisker** [comment]
[test katello]

**Review by ezr-ondrej — APPROVED** [review]
LGTM :+1:
@amirfefer or @laviro any comments? :)

File: webpack/assets/javascripts/react_app/redux/actions/toasts/index.js [hunk 2]
```diff
@@ -1,22 +1,9 @@
...
-};
-
-export const deleteToast = key => ({
-  type: TOASTS_DELETE,
-  payload: { key },
-});
-
-export const clearToasts = () => ({ type: TOASTS_CLEAR });
+export {
+  addToast,
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/redux/actions/toasts/index.js:6]: great idea to pass the reference to the new implementation and show a deprecation warning, I was worried for plugins 👍🏼

File: webpack/assets/javascripts/react_app/components/ToastsList/slice.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import uuidV1 from 'uuid/v1';
+
+const initialState = {};
+const toastsListSlice = createSlice({
+  name: 'toasts',
+  initialState,
+  reducers: {
+    addToast: {
+      reducer: (state, { payload: { key, toast } }) => {
+        state[key] = toast;
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/ToastsList/slice.js:11]: so earlier `initialState` was `Immutable({ messages: {} });` 
and also `state.setIn(['messages', payload.key]....`

and here you've simplified it more, nice!

File: webpack/assets/javascripts/react_app/components/ToastsList/slice.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      },
+      prepare: toast => {
+        const key = toast.key || uuidV1();
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/ToastsList/slice.js:14]: I saw  that `redux-toolkit` is exporting `nanoid`: https://soyoung210.github.io/redux-toolkit/api/other-exports/#nanoid

which does the same thing as the UUID lib, maybe we can drop the usage of `uuid/v1`

**Review by Ron-Lavi — APPROVED** [review]
Nice cleanup @tbrisker !
tested and works well on core and also on plugins,
code-wise seem great, we should consider using [nanoid](https://soyoung210.github.io/redux-toolkit/api/other-exports/#nanoid) which is being exported via `redux-toolkit` but that is for another discussion

**Comment by Ron-Lavi** [comment]
Thanks @tbrisker & @ezr-ondrej !

**Comment by tbrisker** [comment]
Thanks! opened https://github.com/theforeman/foreman/pull/8860 to switch it to nanoid

**Pull request merged**